draft/celery-5-high-level-architecture.rst

```diff
@@ -0,0 +1,1216 @@
+In order to be network resilient we must use retries and circuit breakers on
+all outgoing and incoming connections.
+
+Command Line Interface
+----------------------
+
+Our command line interface is the user interface to all of Celery's
+functionality. It is crucial for us to provide an excellent user experience.
+
+Currently Celery uses :mod:`argparse` with a few custom hacks and workarounds for
+things which are not possible to do with :mod:`argparse`.
+This created some bugs in the past.
+
+Celery 5 will use `Click`_, a modern Python library for creating command line
+programs.
+
+Click's documentation `explains <https://click.palletsprojects.com/en/7.x/why>`_
+why it is a good fit for us:
+
+  There are so many libraries out there for writing command line utilities;
+  why does Click exist?
+
+  This question is easy to answer: because there is not a single command
+  line utility for Python out there which ticks the following boxes:
+
+  * is lazily composable without restrictions
+  * supports implementation of Unix/POSIX command line conventions
+  * supports loading values from environment variables out of the box
+  * supports for prompting of custom values
+  * is fully nestable and composable
+  * works the same in Python 2 and 3
+  * supports file handling out of the box
+  * comes with useful common helpers (getting terminal dimensions,
+    ANSI colors, fetching direct keyboard input, screen clearing,
+    finding config paths, launching apps and editors, etc.)
+
+  There are many alternatives to Click and you can have a look at them if
+  you enjoy them better.  The obvious ones are ``optparse`` and ``argparse``
+  from the standard library.
+
+  Click actually implements its own parsing of arguments and does not use
+  ``optparse`` or ``argparse`` following the ``optparse`` parsing behavior.
+  The reason it's not based on ``argparse`` is that ``argparse`` does not
+  allow proper nesting of commands by design and has some deficiencies when
+  it comes to POSIX compliant argument handling.
+
+  Click is designed to be fun to work with and at the same time not stand in
+  your way.  It's not overly flexible either.  Currently, for instance, it
+  does not allow you to customize the help pages too much. This is intentional
+  because Click is designed to allow you to nest command line utilities.  The
+  idea is that you can have a system that works together with another system by
+  tacking two Click instances together and they will continue working as they
+  should.
+
+  Too much customizability would break this promise.
+
+Click describes it's
+`advantages over argparse <https://click.palletsprojects.com/en/7.x/why/#why-not-argparse>`_
+in its documentation as well:
+
+  Click is internally based on optparse instead of argparse.  This however
+  is an implementation detail that a user does not have to be concerned
+  with.  The reason however Click is not using argparse is that it has some
```

Comment:
I think this is repeated [here](https://github.com/celery/ceps/pull/19/files#diff-ba64d98ded44d16b0a7cb52b3d31a59aR713).

Reply:
This is a quote from their documentation.

Reply:
Is this resolved?

Reply:
@georgepsarakis ping?